libraries/stdlib/js/src/kotlin/collections/ArrayList.kt

```diff
@@ -82,10 +83,11 @@ public actual open class ArrayList<E> internal constructor(private var array: Ar
 
         if (index == size) return addAll(elements)
         if (elements.isEmpty()) return false
-        when (index) {
-            size -> return addAll(elements)
-            0 -> array = elements.toTypedArray<Any?>() + array
-            else -> array = array.copyOfRange(0, index).asDynamic().concat(elements.toTypedArray<Any?>(), array.copyOfRange(index, size))
+
+        var i = index
+        elements.forEach { element ->
+            array.asDynamic().splice(i, 0, element)
```

Comment:
Moving array tail on each element insertion would result in quadratic complexity.